nixos/nix-ld: Support 32-bit portability

KoviRobi opened the pull request:
This now supports arbitrary `nix-ld`

Notes for reviewers:
- I could not find a natural way to select the 32-bit equivalent of a 64-bit system, so no `support32Bit` currently.
- Defaults for `libraries` do now get selected from the corresponding `pkgs`
  - This has one remaining design point, see also https://discourse.nixos.org/t/nix-ld-32bit-merging-defaults-to-generated-attrsets/63433

Cross-referencing
- https://github.com/Mic92/nix-ld/issues/28
- https://github.com/nix-community/nix-ld-rs/issues/67

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

zhaofengli reviewed (commented):
This is a good start.

Perhaps instead of special variants, this could be structured more generically as `programs.nix-ld.systems.<system>` where each system attribute is a submodule containing `{ package = ...; ldso = ...; libraries = ...; }`. The existing `package` and `libraries` options are passed down to `programs.nix-ld.systems.${pkgs.system}`. This should make it more composable and support prebuilt binaries for `boot.binfmt.emulatedSystems`.

In this scheme, we can add `support32Bit` which would be a convenience shortcut that determines the corresponding 32-bit `system` (e.g., `armv7l-linux`), the package set (e.g., `pkgsCross.armv7l-hf-multiplatform`), and the default `libraries` to insert.

KoviRobi commented:
Thank you, I think I understand what you are saying, I'll have to try it out. I just haven't had the time/energy to do so yet, but it's not abandoned. And yes I think it'll help with non-Intel/AMD architectures e.g. https://github.com/Mic92/nix-ld/issues/27

KoviRobi commented:
@zhaofengli I'm sorry I'm struggling to figure out how to do this, I keep running into infinite recursion.

For reference, the `builtins.toJSON` line gives the following
```console
$ nix build .#test
$ jq < result/etc/foo 
{
  "environment": {
    "ldso": "/nix/store/s76k7qxy3jj6giv2fzmlx5gycdc6ayy7-nix-ld-1.2.3/libexec/nix-ld",
    "pathsToLink": [
      "/share/nix-ld-x86_64-linux"
    ],
    "systemPackages": [
      "/nix/store/46wx8bbsdfb8r21wj6ddb85j4jkjzxc4-ld-library-path"
    ],
    "variables": {
      "NIX_LD_LIBRARY_PATH_x86_64_linux": "/run/current-system/sw/share/nix-ld-x86_64-linux/lib",
      "NIX_LD_x86_64_linux": "/run/current-system/sw/share/nix-ld-x86_64-linux/lib/ld.so"
    }
  },
  "programs": {
    "nix-ld": {
      "systems": {
        "x86_64-linux": {
          "libraries": [
            "/nix/store/rc18a8k50zmrif61250sfidkqvlg41ln-zlib-1.3.1",
            "/nix/store/w6ggciw9mj6r6w332dclhibwmyyv1v67-zstd-1.5.6-bin",
            "/nix/store/3s3rjkl3mx05wp0lmxgwkaqhbz9sy6kk-gcc-13.3.0",
            "/nix/store/did64gsb4sn6xvylw6cswa7697avdrsb-curl-8.8.0-bin",
            "/nix/store/ppfzl2wqzs7ljfj0dxypcvrhr9x6ka3f-openssl-3.0.14-bin",
            "/nix/store/f6vjq8dl0z6x1k7sxvzwnisbncbafnjy-attr-2.5.2-bin",
            "/nix/store/adjqi4kf1k736g7fz23hdqsk39zyjzf8-libssh-0.10.6",
            "/nix/store/wr6x94hzs5ggmpql0qx5j8q2rpfrhm7q-bzip2-1.0.8-bin",
            "/nix/store/iyaxbb5c472akzpbsn3wpmfmh8xzw7k4-libxml2-2.12.7-bin",
            "/nix/store/9xw9hxlmxkjcv02x9d4wrj3sl8gwmkl9-acl-2.3.2-bin",
            "/nix/store/8kd6icdk2025jmvhvxhi6rc3ahhj2qc4-libsodium-1.0.20",
            "/nix/store/kwa0akdagjw0hmv3mam65gmz8f1v8c8k-util-linux-2.39.4-bin",
            "/nix/store/0rlbqq6p9wwb7wbn5vnpck40c76czid3-xz-5.6.2-bin",
            "/nix/store/sbqf71kqhlgaff1ajd8zak09g4wn8ymd-systemd-255.6"
          ]
        }
      }
    }
  }
}
```
but if I uncomment the `# recursive` at the bottom of the WIP comment `nix-ld.nix`, then I get an infinite recursion, even if I remove the `programs.nix-ld.systems.${system}.libraries` bit.

KoviRobi commented:
Ok I have taken a bigger break than I intended to, but coming back to this, I figured out the issue, it was because I was doing

```nix
{ pkgs, lib, config, ... }:
let
  inherit (lib) mapAttrsToList mergeAttrsList;
  cfg = config.programs.nix-ld.systems;
in
{
  config =
    mergeAttrsList
      (mapAttrsToList
        (system: cfg: {
          environment.variables = ...;
          # ...
        })
        cfg);
}
```

which creates a dynamic lazy top-level set. If I pushed the dynamic-ness inside, so that it can never produce a value for `config.programs.nix-ld`, then it removes the infinite recursion, i.e.

```nix
{ pkgs, lib, config, ... }:
let
  inherit (lib) mapAttrsToList mergeAttrsList;
  cfg = config.programs.nix-ld.systems;
in
{
  config = {
    environment.variables = mergeAttrsList (mapAttrsToList ...);
  };
}
```

TODO: Should I fix existing tests to not emit a warning due to `mkRenamedOptionModule`? It's been useful to not do so currently to check backwards compatibility.

Commit:
nixos/nix-ld: Add nix-ld 32-bit variant

I have run `nix build .\#nix-ld.passthru.tests.nix-ld{,-32bit}` and
they work.

The new test (run on an x86_64-linux install, if that matters) shows
nix-ld-32bit working.

KoviRobi commented:
Also forgot to tag maintainer @Mic92

KoviRobi commented:
After using it in practice, one issue is that the library defaults don't get merged, but rather are overridden. Anyone with more experience of NixOS modules can help me?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-ld-32bit-merging-defaults-to-generated-attrsets/63433/1

Mic92 commented on nixos/modules/programs/nix-ld.nix:
```diff
@@ -5,62 +5,146 @@
+      system = pkgs.system or ''''${pkgs.system}'';
+    in
+    [
+      (mkRenamedOptionModule
```
I would prefer to make `"programs" "nix-ld" "enable"` an alias instead of deprecating it, because it makes it easier to enable nix-ld across architectures.

KoviRobi replied:
Done, apparently `mkAliasOptionModuleMD` is itself no longer needed and just an alias for `mkAliasOptionModule`

KoviRobi replied:
Thank you, this does resolve the issue about the deprecation warning, and I do agree it makes it more portable. Now the remaining question is about the defaults

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/32-bit-doesnt-support-on-nixos/68483/8

depeo commented:
This would be a great addition, I have had a lot of issues with 32-bit libraries, when trying to sandbox Lutris/Bottles.
Is this something that will be added to NixOS?

KoviRobi commented:
I've been using it personally for a while, the remaining awkward thing is my comment above
https://github.com/NixOS/nixpkgs/pull/326948#issuecomment-2744033461
from my viewpoint. But also I haven't got any more ideas, or thought about it for ages

Edit: But as an added bonus I have an aarch64 laptop so can also play around with that
